app/models/user.rb

```diff
@@ -18,6 +19,38 @@ class User < ApplicationRecord
                          wrong_length: 'must be between 3 and 20 characters'
                        }
 
+  class << self
+    def from_omniauth(auth)
+      username = auth.info.email.split('@').first
+
+      find_or_create_by(
+        provider: auth.provider,
+        uid: auth.uid,
+        params: new_user_params(auth, username)
+      )
+    end
+
+    private
+
+    def can_use_username?(username, email)
+      !find_by(username:, email:).blank?
```

Comment:
it looks like this could work too
```suggestion
      !exists?(username:, email:)
```